<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview exam questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

<div class="alert alert-warning" role="alert" style="margin: 10px">
<p style="font-weight:bold">WARNING</p>

<p>If using <a href="https://try.jupyter.org">try.jupyter.org</a> do not rely on the server for anything you want to last - your server will be <span style="font-weight:bold">deleted after 10 minutes of inactivity</span>. Save often and rember download notebook when you step away (you can always re-upload and start again)!</p>
</div>


## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we'll look forward to following up!

# Exam Questions

## 1. Algo + Data Structures

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

In [23]:
def fibonacci(n): 
   
    
    if n==1 or n == 2: 
        return 1
    
    else: 
        return fibonacci(n-1)+fibonacci(n-2) 
  

  
fibonacci(9)

34

#### Q 1.1.2
What's the complexity of your implementation?

This implementation is a recursive one and hence the time complexity for n is exponential. 
C(n) = C(n-1) + C(n-2)
Since it is exponential the complexity can be approximated to $O(2^n)$
Also there is a repetitive calculations happening as each time f(n-1) and f(n-2) will calculate the underlying fibonacci numbers again and again. Hence space complexity is high.


#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

In [29]:
def fibonacci2(n):  
    f_1 = 1
    f_2 = 1
    if n == 1 or n ==2: 
        return f_1 
    
    else: 
        for i in range(3,n+1): 
            f_n = f_1 + f_2 
            f_1 = f_2 
            f_2 = f_n 
        return f_2 
  
  
print(fibonacci2(9)) 

34


#### Q 1.1.4
What's the complexity of your implementation?


This is a O(n) complex implementation. Also since we are only storing the last two fibonaccis each time it is a space optimized solution as well.

#### Q 1.1.5
What are some examples of optimizations that could improve computational performance?


The repeated work done in the first fibonacci function can be avoided and optimized by storing the already calculated fibonacci numbers in an array. This will reduce the space complexity many folds. 



Also if we can calculate the generalized formula for fibonacci(n) mathematically then our solution will be O(1)

### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


#### Q 1.2.2
What is the complexity of your implementation?

#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

#### Q 1.2.4
What's the complexity of this implementation?


#### Q 1.2.5 
What are some examples of optimizations that could improve computational performance?


## 2. Prob + Stats

### Q 2.1: Finding $\pi$ in a random uniform?
<img src=https://www.epicurus.com/food/recipes/wp-content/uploads/2015/03/Pi-Day.jpg width="480">

Given a uniform random generator $[0,1)$ (e.g., use your language's standard libary to generate random value), write a a function `compute_pi` to compute [$\pi$](https://en.wikipedia.org/wiki/Pi).

In [43]:
def compute_pi():
    import random

    quarter=0
    square=0

    for i in range(1000000):

        x= random.uniform(0, 1) 
        y= random.uniform(0, 1) 

        eucl_dist = x**2 + y**2


        if eucl_dist<= 1: 
            quarter+= 1

        square+= 1

        pi = 4* quarter/square

    print("Pi approx=", pi)    
    
compute_pi()

Pi approx= 3.141096


<img src="compute_pi.png">

### Q 2.2: Making a 6-side die roll a 7?

Using a single 6-side die, how can you generate a random number between 1 - 7?

### Q 2.3: Is normality uniform?

<img src=https://rednaxela1618.files.wordpress.com/2014/06/uniformnormal.png width="480">


Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.4: Should you pay or should you go?

![coin flip](https://lh5.ggpht.com/iwD6MnHeHVAXNBgrO7r4N9MQxxYi6wT9vb0Mqu905zTnNlBciONAA98BqafyjzC06Q=w300)

Let’s say we play a game where I keep flipping a coin until I get heads. If the first time I get heads is on the nth coin, then I pay you $2^{(n-1)}$ US dollars. How much would you pay me to play this game? Explain.

### Q 2.5: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID, what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?

### Q 2.6: Pick your prize
<img src=https://miro.medium.com/max/1100/1*m5b3O9sE68UCXjLw5oxy2g.png width="480">

A prize is placed at random behind one of three doors and you are asked to pick a door. To be concrete, say you always pick door 1. Now the game host chooses one of door 2 or 3, opens it and shows you that it is empty. They then give you the option to keep your picked door or switch to the unopened door. Should you stay or switch if you want to maximize your probability of winning the prize?

## 3 Conceptual ML

### Q 3.1 Why study gradient boosting or neural networks?

Consider a regression setting where $X \in \mathbb{R}^p$ and $Y \in \mathbb{R}$. The goal is to come up with a function $f(X): \mathbb{R}^p \rightarrow \mathbb{R}$ that minimizes the squared-error loss $(Y - f(X))^2$. Since X, Y are random variables, we seek to minimize the expectation of the squared error loss as follows
\begin{equation}
EPE(f) = \mathbb{E}\left[(Y-f(X)^2\right]
\end{equation}
where EPE stands for expected prediction error. One can show that minimizing the expected prediction error leads to the following _regression function_
\begin{equation}
f(x) = \mathbb{E}\left[Y|X=x\right]
\end{equation}

The goal of any method is to approximate the regression function above, which we denote as $\hat{f}(x)$. For example, linear regression explicitly assumes that the regression function is approximately linear in its arguments, i.e. $\hat{f}(x) = x^T\beta$ while a neural network provides a nonlinear approximation of the regression function. 

The simplest of all these methods is [k-nearest neighbors](https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm). Given $x$ and some neighbourhood of $k$ points $N_k(x)$, $\hat{f}(x)$ is simply the average of all $y_i|x_i \in N_k(x)$.  Let $N$ denote the training sample size. Under mild regularity conditions on the joint probability distribution $Pr(X, Y)$, one can show that as $N \rightarrow \infty$, $k \rightarrow \infty$ such that $k/N \rightarrow 0$, then $\hat{f}(x) \rightarrow f(x)$ where $\rightarrow$ means approaches or goes to. In other words, the k-nearest neighbors algorithm converges to the ideal solution as both the training sample size and number of neighbors increase to infinity.

Now given this _universal approximator_, why look any further and research other methods? Please share your thoughts.


### Q 3.2 Model Selection and Assesment

Consider a multiclass classification problem with a large number of features $p >> N$, for e.g $p=10000, N=100$ The task is threefold
1. Find a "good" subset of features that show strong _univariate_ correlation with class labels
2. Using the "good" subset, build a multi class classifier
3. Estimate the generalization error of the final model

Given this dataset, outline your approach and please be sure to cover the following
- Data splitting
- Model Selection: either estimating the performance of different classifiers or the same classifier with different hyperparameters
- Model Assessment: having chosen a classifier, estimating the generalization error

Assume all features are numerical, the dataset contains no NULLS, outliers, etc. and doesn't require any preprocessing.



####  a) Finding "good" subset of features that show strong univariate correlation with class labels:

1. We can build a baseline model with all of the data and incrementally select features that improve model performance. This is a greedy algo approach were we iterate p times and build a model and check the performance. 

2. We can use Spectral Value Decomposition or Dimensionality reduction to project the data into a smaller dimension. Although this will make the results derived from the model uninterpretable.

3. We can go for a naive selection of features were we choose top k features based on mutual information/correlation with Y. This can be done by building a simple decision tree and check the feature importances. But as the sample size is large and number of features is less decision tree is not a very good method for this as it tends to overfit in such cases.

4. We can check for correlation and collinearity within the features and remove them as it might hinder model performance. 

5. The best way would be to do feature selection for each algorithm and incorporate any of the 1st three methods into the iterative approach detailed below. 


#### b) Using the "good" subset, build a multi class classifier:

1. We can choose and build a baseline model with the above identified features as mentioned above, given such a dataset I would go for logistic regression as it performs well with smaller sample size.

2. Since the dataset is small we can train the model with 80-85% of the data and use crossvalidation to validate and get a better model.

3. We can test each model with the remaining 10-15% test data for comparison.

4. Based on the problem we are dealing we can choose an appropriate evaluation metric for comparison. eg Precision, Recall, AUC etc.

5. Run with different hyper parameters of the model to obtain a good model.

5. We can repeat the steps 2,3,4,5 for different models like SVM, Random Forest, Gradient Boosting etc. to get the best model with the given data. Ideally we are doing algorithm selection feature selection and hyper parameter selection.

#### c) Estimate the generalization error of the final model

Generalization error consists of bias variance and irreducible error. The goals of modeling is generally not to hit some target level of bias or variance instead, we want to reduce the prediction error.The total model error is determined by the amount of bias or variance. Thus, controlling bias and variance in essence helps us control the error.
Since the data size is less we will have a lot of variance factor in our generalization error. If we add more data we can reduce that but the value of this variance reduction might not be proportional to the costs associated with increased data.

Increasing model complexity will increase bias in our data but the proportion is not quantifiable. We have done our model selection in a comprehensive way and so we need to tradeoff between these two factors.

Ultimately we need to tradeoff between these two factors to obtain a good model. The best performing model on out of sample test data will be best one in terms of its bias and variance. For different algorithms we have used we can check the plot between training error and prediction error to obtain an optimal bias variance trade off. 